kv/kvserver: TestRangefeedCheckpointsRecoverFromLeaseExpiration failed

kv/kvserver.TestRangefeedCheckpointsRecoverFromLeaseExpiration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15268321?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15268321?buildTab=artifacts#/) on release-24.1 @ [556680db129ea635342767fdd916628409071ef1](https://github.com/cockroachdb/cockroach/commits/556680db129ea635342767fdd916628409071ef1):


```
=== RUN   TestRangefeedCheckpointsRecoverFromLeaseExpiration
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/37ddc8e37e5448832e32de57f39c38ec/logTestRangefeedCheckpointsRecoverFromLeaseExpiration544955870
    test_log_scope.go:81: use -show-logs to present logs inline
    replica_rangefeed_test.go:1517: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_rangefeed_test.go:1517
        	Error:      	Received unexpected error:
        	            	liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715719335.242344971,0); actual is liveness(nid:2 epo:1 exp:1715719334.132890321,0); is the node still live?
        	            	(1) liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715719335.242344971,0); actual is liveness(nid:2 epo:1 exp:1715719334.132890321,0); is the node still live?
        	            	Error types: (1) *liveness.ErrEpochCondFailed
        	Test:       	TestRangefeedCheckpointsRecoverFromLeaseExpiration
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/37ddc8e37e5448832e32de57f39c38ec/logTestRangefeedCheckpointsRecoverFromLeaseExpiration544955870
--- FAIL: TestRangefeedCheckpointsRecoverFromLeaseExpiration (1.64s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #123551 kv/kvserver: TestRangefeedCheckpointsRecoverFromLeaseExpiration failed [C-bug C-test-failure O-robot T-kv branch-master release-blocker skipped-test]
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangefeedCheckpointsRecoverFromLeaseExpiration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38746

kv/kvserver.TestRangefeedCheckpointsRecoverFromLeaseExpiration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15268962?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15268962?buildTab=artifacts#/) on release-24.1 @ [a3f808338f097ce9f2b183ad65311a96d8fa5766](https://github.com/cockroachdb/cockroach/commits/a3f808338f097ce9f2b183ad65311a96d8fa5766):


```
=== RUN   TestRangefeedCheckpointsRecoverFromLeaseExpiration
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/37ddc8e37e5448832e32de57f39c38ec/logTestRangefeedCheckpointsRecoverFromLeaseExpiration3693211046
    test_log_scope.go:81: use -show-logs to present logs inline
    replica_rangefeed_test.go:1517: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_rangefeed_test.go:1517
        	Error:      	Received unexpected error:
        	            	liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715719915.062259685,0); actual is liveness(nid:2 epo:1 exp:1715719913.950302220,0); is the node still live?
        	            	(1) liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715719915.062259685,0); actual is liveness(nid:2 epo:1 exp:1715719913.950302220,0); is the node still live?
        	            	Error types: (1) *liveness.ErrEpochCondFailed
        	Test:       	TestRangefeedCheckpointsRecoverFromLeaseExpiration
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/37ddc8e37e5448832e32de57f39c38ec/logTestRangefeedCheckpointsRecoverFromLeaseExpiration3693211046
--- FAIL: TestRangefeedCheckpointsRecoverFromLeaseExpiration (1.66s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #123551 kv/kvserver: TestRangefeedCheckpointsRecoverFromLeaseExpiration failed [C-bug C-test-failure O-robot T-kv branch-master release-blocker skipped-test]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangefeedCheckpointsRecoverFromLeaseExpiration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

kv/kvserver.TestRangefeedCheckpointsRecoverFromLeaseExpiration [failed](https://tanzanite.cluster.engflow.com/invocations/default/059e1895-617e-421f-9183-9b5b51c959a2?testReportRun=17&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-24.1 @ [6e68cb2b3a4fb7f02cfafcccc5f81d6356de171e](https://github.com/cockroachdb/cockroach/commits/6e68cb2b3a4fb7f02cfafcccc5f81d6356de171e):


```
=== RUN   TestRangefeedCheckpointsRecoverFromLeaseExpiration
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRangefeedCheckpointsRecoverFromLeaseExpiration460166722
    test_log_scope.go:81: use -show-logs to present logs inline
    replica_rangefeed_test.go:1517: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_rangefeed_test.go:1517
        	Error:      	Received unexpected error:
        	            	liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715758759.283332402,0); actual is liveness(nid:2 epo:1 exp:1715758758.182541957,0); is the node still live?
        	            	(1) liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715758759.283332402,0); actual is liveness(nid:2 epo:1 exp:1715758758.182541957,0); is the node still live?
        	            	Error types: (1) *liveness.ErrEpochCondFailed
        	Test:       	TestRangefeedCheckpointsRecoverFromLeaseExpiration
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestRangefeedCheckpointsRecoverFromLeaseExpiration460166722
--- FAIL: TestRangefeedCheckpointsRecoverFromLeaseExpiration (1.63s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=17</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #123551 kv/kvserver: TestRangefeedCheckpointsRecoverFromLeaseExpiration failed [C-bug C-test-failure O-robot T-kv branch-master release-blocker skipped-test]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangefeedCheckpointsRecoverFromLeaseExpiration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

kv/kvserver.TestRangefeedCheckpointsRecoverFromLeaseExpiration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15276339?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15276339?buildTab=artifacts#/) on release-24.1 @ [28735b05c5a8caae045a92939a5dc233d0c3915f](https://github.com/cockroachdb/cockroach/commits/28735b05c5a8caae045a92939a5dc233d0c3915f):


```
=== RUN   TestRangefeedCheckpointsRecoverFromLeaseExpiration
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/37ddc8e37e5448832e32de57f39c38ec/logTestRangefeedCheckpointsRecoverFromLeaseExpiration2079952082
    test_log_scope.go:81: use -show-logs to present logs inline
    replica_rangefeed_test.go:1517: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_rangefeed_test.go:1517
        	Error:      	Received unexpected error:
        	            	liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715768670.279313075,0); actual is liveness(nid:2 epo:1 exp:1715768669.115703473,0); is the node still live?
        	            	(1) liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715768670.279313075,0); actual is liveness(nid:2 epo:1 exp:1715768669.115703473,0); is the node still live?
        	            	Error types: (1) *liveness.ErrEpochCondFailed
        	Test:       	TestRangefeedCheckpointsRecoverFromLeaseExpiration
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/37ddc8e37e5448832e32de57f39c38ec/logTestRangefeedCheckpointsRecoverFromLeaseExpiration2079952082
--- FAIL: TestRangefeedCheckpointsRecoverFromLeaseExpiration (1.99s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #123551 kv/kvserver: TestRangefeedCheckpointsRecoverFromLeaseExpiration failed [C-bug C-test-failure O-robot T-kv branch-master release-blocker skipped-test]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangefeedCheckpointsRecoverFromLeaseExpiration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

kv/kvserver.TestRangefeedCheckpointsRecoverFromLeaseExpiration [failed](https://tanzanite.cluster.engflow.com/invocations/default/52bcdd6b-c417-4f2e-9412-1c0f3347fdff?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-24.1 @ [6e68cb2b3a4fb7f02cfafcccc5f81d6356de171e](https://github.com/cockroachdb/cockroach/commits/6e68cb2b3a4fb7f02cfafcccc5f81d6356de171e):


```
=== RUN   TestRangefeedCheckpointsRecoverFromLeaseExpiration
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRangefeedCheckpointsRecoverFromLeaseExpiration3831727626
    test_log_scope.go:81: use -show-logs to present logs inline
    replica_rangefeed_test.go:1517: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_rangefeed_test.go:1517
        	Error:      	Received unexpected error:
        	            	liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715770804.597098206,0); actual is liveness(nid:2 epo:1 exp:1715770802.133394759,0); is the node still live?
        	            	(1) liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1715770804.597098206,0); actual is liveness(nid:2 epo:1 exp:1715770802.133394759,0); is the node still live?
        	            	Error types: (1) *liveness.ErrEpochCondFailed
        	Test:       	TestRangefeedCheckpointsRecoverFromLeaseExpiration
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestRangefeedCheckpointsRecoverFromLeaseExpiration3831727626
--- FAIL: TestRangefeedCheckpointsRecoverFromLeaseExpiration (5.66s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #123551 kv/kvserver: TestRangefeedCheckpointsRecoverFromLeaseExpiration failed [C-bug C-test-failure O-robot T-kv branch-master release-blocker skipped-test]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangefeedCheckpointsRecoverFromLeaseExpiration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/kv/kvserver/replica_rangefeed_test.go
+++ b/pkg/kv/kvserver/replica_rangefeed_test.go
@@ -1521,8 +1521,6 @@ func TestRangefeedCheckpointsRecoverFromLeaseExpiration(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.WithIssue(t, 123551)
-
 	ctx := context.Background()
 	var scratchRangeID int64 // accessed atomically
 	// nudgeSeen will be set if a request filter sees the signature of the
@@ -1663,13 +1661,25 @@ func TestRangefeedCheckpointsRecoverFromLeaseExpiration(t *testing.T) {
 	// Expire the lease. Given that the Raft leadership is on n2, only n2 will be
 	// eligible to acquire a new lease.
 	log.Infof(ctx, "test expiring lease")
-	nl := n2.NodeLiveness().(*liveness.NodeLiveness)
-	resumeHeartbeats := nl.PauseAllHeartbeatsForTest()
-	n2Liveness, ok := nl.Self()
+	nl2 := n2.NodeLiveness().(*liveness.NodeLiveness)
+	resumeHeartbeats := nl2.PauseAllHeartbeatsForTest()
+	n2Liveness, ok := nl2.Self()
 	require.True(t, ok)
-	manualClock.Increment(n2Liveness.Expiration.ToTimestamp().Add(1, 0).WallTime - manualClock.UnixNano())
+	manualClock.Increment(max(firstLease.MinExpiration.WallTime, n2Liveness.Expiration.ToTimestamp().
+		Add(1, 0).WallTime) - manualClock.UnixNano())
 	atomic.StoreInt64(&rejectExtraneousRequests, 1)
-	// Ask another node to increment n2's liveness record.
+
+	// Ask another node to increment n2's liveness record, but first, wait until
+	// n1's liveness state is the same as n2's. Otherwise, the epoch below might
+	// get rejected because of mismatching liveness records.
+	testutils.SucceedsSoon(t, func() error {
+		nl1 := n1.NodeLiveness().(*liveness.NodeLiveness)
+		n2LivenessFromN1, _ := nl1.GetLiveness(n2.NodeID())
+		if n2Liveness != n2LivenessFromN1.Liveness {
+			return errors.Errorf("waiting for node 2 liveness to converge on both nodes 1 and 2")
+		}
+		return nil
+	})
 	require.NoError(t, n1.NodeLiveness().(*liveness.NodeLiveness).IncrementEpoch(ctx, n2Liveness))
 	resumeHeartbeats()
 

